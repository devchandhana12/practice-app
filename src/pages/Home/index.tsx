import React from "react";
import axios from "axios";
import { FaStar } from "react-icons/fa";
import { AuthContext } from "../../contexts/AuthContext";
import Header from "../../components/Header";
import Footer from "../../components/Footer";
import CardComp from "../../components/Card";
import SubMenu from "../../components/SubMenu";
import ButtonComp from "../../components/Button";
interface Product {
  id: number;
  title: string;
  image: string;
  price: string;
  rating: {
    rate: string;
    count: string;
  };
}

const Home = () => {
  const [data, setData] = React.useState<Product[] | undefined>(undefined);
  const [categories, setCategories] = React.useState([]);
  const [cartValue, setCartValue] = React.useState<number>(0);
  React.useEffect(() => {
    (async () => {
      await axios
        .get("https://fakestoreapi.com/products")
        .then((res) => setData(res.data))
        .catch((error) => console.log(error));
      await axios
        .get("https://fakestoreapi.com/products/categories")
        .then((res) => setCategories(res.data))
        .catch((e) => console.log(e));
    })();
  }, []);

  const handleCart = () => {
    alert(cartValue);
    setCartValue(cartValue + 1);
  };
  const { user } = React.useContext(AuthContext);
  console.log(user);
  return (
    <div>
      <Header user={user} />

      <div className="d-flex">
        <SubMenu data={categories} />
        <div
          className="d-flex"
          style={{
            flexWrap: "wrap",
            justifyContent: "center",
            alignItems: "center",
          }}
        >
          {data?.map((item) => (
            <div key={item?.id}>
              <CardComp
                style={{
                  height: 450,
                  width: 350,
                  margin: 8,
                  display: "flex",
                  flexDirection: "column",
                  alignItems: "center",
                }}
              >
                <img
                  src={item.image}
                  className="img-fluid"
                  alt={item.title}
                  style={{ height: 300, width: 400 }}
                />
                <p className="text-center fs-6 fst-italic mt-3 fw-bolder">
                  {item?.title?.toString().substring(0, 30) + "..."}
                </p>
                <div>
                  <div className="d-flex align-items-center">
                    <FaStar style={{ color: "#40F8FF" }} />
                    <div className="mt-4 m-2 d-flex">
                      <p className="fs-6 fw-normal fst-italic">
                        {item?.rating?.rate}
                      </p>
                      <p style={{ color: "#63686E" }} className="fw-bold">
                        ({item?.rating?.count})
                      </p>
                    </div>
                    <p className="d-flex align-items-center mt-2 m-2">
                      Price :{" "}
                      <p className="mt-3 ms-1" style={{ color: "#40F8FF" }}>
                        ${Math.round(parseInt(item?.price))}
                      </p>
                    </p>
                    <div className="ms-1">
                      <ButtonComp
                        label="-"
                        variant="danger"
                        color="black"
                        onClick={() => {
                          if (cartValue > 0) {
                            setCartValue(cartValue - 1);
                          }
                        }}
                      />
                      <ButtonComp label={cartValue} disabled color="black" />
                      <ButtonComp
                        label="+"
                        variant="success"
                        color="black"
                        type="button"
                        onClick={handleCart}
                      />
                    </div>
                  </div>
                </div>
              </CardComp>
            </div>
          ))}
        </div>
      </div>
      <Footer />
    </div>
  );
};
export default Home;
