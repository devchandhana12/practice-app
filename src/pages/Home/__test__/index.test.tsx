/* eslint-disable testing-library/no-wait-for-multiple-assertions */
import Home from "..";
import { render, cleanup, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import fetchMock from "jest-fetch-mock";
afterEach(cleanup);

it("should match snapshot", () => {
  const { asFragment } = render(<Home />);
  expect(asFragment()).toMatchSnapshot();
});

jest.mock("axios");
const mockdata = [
  {
    id: 1,
    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
    price: 109.95,
    description:
      "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
    category: "men's clothing",
    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
    rating: {
      rate: 3.9,
      count: 120,
    },
  },
];
//
test("renders products when API call succeeds", async () => {
  // Mock the API response
  fetchMock.mockResponseOnce(JSON.stringify(mockdata), { status: 200 });

  // Render the Home component
  render(<Home />);

  // Wait for the loading element to not be present
  await waitFor(() => {
    expect(screen.queryByTestId("loading_text")).not.toBeInTheDocument();
  });

  // Your other assertions for the successful rendering of products
  expect(screen.getByTestId("cards")).toBeInTheDocument();
});
