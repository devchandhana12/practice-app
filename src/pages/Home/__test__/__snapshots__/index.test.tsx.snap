// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="d-flex align-items-center justify-content-around shadow-lg navbar navbar-expand navbar-light"
      style="height: 70px; background-color: rgb(53, 89, 224);"
    >
      <a
        class="text-white navbar-brand"
        href="#home"
        style="font-weight: bolder; font-style: italic;"
      >
        Fakestore
      </a>
      <div>
        <div
          class="d-flex align-items-center m-5 p-2 rounded"
          style="background-color: rgb(48, 129, 208);"
        >
          <input
            placeholder="Search"
            style="background-color: transparent; outline: none; color: white; width: 450px;"
            value=""
          />
          <svg
            fill="currentColor"
            height="22"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="22"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
        </div>
      </div>
      <div
        class="d-flex align-items-center "
      >
        <a
          data-testid="login"
          href="/login"
          style="color: white; font-size: 16px; font-weight: bolder; font-style: italic;"
        >
          Login
        </a>
        <button
          class="fs-6 fw-bold fst-italic btn btn-danger"
          style="margin-right: 20px; margin-left: 20px;"
        >
          Logout
        </button>
      </div>
    </nav>
    <div
      class="d-flex justify-conten-around"
    >
      <div
        class=""
        style="height: 100vh; left: 0px; z-index: 1; width: 400px;"
      />
      <div
        class="d-flex"
        style="flex-wrap: wrap; justify-content: center; align-items: center;"
      >
        <div>
           
        </div>
      </div>
    </div>
    <div>
      im a footer to your site
    </div>
  </div>
</DocumentFragment>
`;
